[export] handle constant aliasing for export

pianpwk opened the pull request:
Summary: Currently export will [error out](https://github.com/pytorch/pytorch/blob/2b5ae2611e22d992565f202df9267fe66469efaa/torch/export/_trace.py#L477) if a constant is aliased. This PR supports this by modifying ConstantAttrMap to map constants to a list of FQNs instead of a single FQN, populating the ExportedProgram constants dict to contain multiple entries to the same constant.

Test Plan: added test case in test_export.py

Differential Revision: D56955654

ydwu4 commented on torch/_export/non_strict_utils.py:
```diff
@@ -340,25 +340,13 @@ def _gather_constant_attrs(m: torch.nn.Module) -> ConstantAttrMap:
 
     def inner(m: torch.nn.Module, prefix_atoms: List[str], constants):
         for k, v in m.__dict__.items():
-            if isinstance(
```
why we want to remove FakeScriptObject?

pianpwk replied:
Oh sorry, this was an oversight.

ydwu4 commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -136,7 +145,7 @@ def lift_constants_pass(
                 # of this get_attr to point to the already-existing placeholder
                 # node.
-                const_placeholder_node = lifted_objs[constant_val]
+                const_placeholder_node = lifted_objs[constant_val][0]
```
probably should change the pattern of constant_fqns.get then [0] to just `get_first_fqn` or something similar to avoid the redundancy.

ydwu4 commented on test/export/test_export.py:
```diff
@@ -4590,6 +4590,40 @@ def forward(self, x):
         }
         export(f, (inputs,), dynamic_shapes=dynamic_shapes)
 
+    def test_constant_aliasing(self):
```
Can you do a rebase on https://github.com/pytorch/pytorch/pull/124978 and add three non-strict mode tests for script objects: 1. one as module attributes, 2. one as input and 3. when it's fakified as FakeScriptObject? Can use the TensorQueue or Foo. Examples can be found in test_torchbind.py.

pianpwk replied:
Will do, thanks

ydwu4 reviewed (commented):
Overall looks good to me but might need to add a few more test cases to guard against regression.

pianpwk committed:
[export] handle constant aliasing for export (#125509)

Summary:

Currently export will error out if a constant is aliased. This PR supports this by modifying ConstantAttrMap to map constants to a list of FQNs instead of a single FQN, populating the ExportedProgram constants dict to contain multiple entries to the same constant.

Test Plan: added test case in test_export.py

Reviewed By: ydwu4, angelayi

Differential Revision: D56955654
